src/Microsoft.SqlTools.ServiceLayer/QueryExecution/SerializationService.cs

```diff
@@ -40,93 +41,106 @@ public override void InitializeService(IProtocolEndpoint serviceHost)
+            return Task.FromResult<bool>(true);
+        }
+
+        internal async Task RunSerializeStartRequest(SerializeDataStartRequestParams serializeParams, RequestContext<SerializeDataResult> requestContext)
         {
             try
             {
+                // Verify we have sensible inputs and there isn't a task running for this file already
                 Validate.IsNotNull(nameof(serializeParams), serializeParams);
                 Validate.IsNotNullOrWhitespaceString("FilePath", serializeParams.FilePath);
 
                 DataSerializer serializer = null;
-                bool hasSerializer = inProgressSerializations.TryGetValue(serializeParams.FilePath, out serializer);
-                if (hasSerializer)
+                if (inProgressSerializations.TryGetValue(serializeParams.FilePath, out serializer))
                 {
+                    // Cannot proceed as there is an in progress serialization happening
                     throw new Exception(SR.SerializationServiceRequestInProgress(serializeParams.FilePath));
                 }
-                
+
+                // Create a new serializer, save for future calls if needed, and write the request out
                 serializer = new DataSerializer(serializeParams);
                 if (!serializeParams.IsLastBatch)
                 {
                     inProgressSerializations.AddOrUpdate(serializer.FilePath, serializer, (key, old) => serializer);
                 }
-                Func<Task<SerializeDataResult>> writeData = () =>
-                {
-                    return Task.Factory.StartNew(() =>
-                    {
-                        var result = serializer.ProcessRequest(serializeParams);
-                        return result;
-                    });
-                };
-                await HandleRequest(writeData, requestContext, "HandleSerializeStartRequest");
+                
+                Logger.Write(TraceEventType.Verbose, "HandleSerializeStartRequest");
+                SerializeDataResult result = serializer.ProcessRequest(serializeParams);
+                await requestContext.SendResult(result);
             }
             catch (Exception ex)
             {
-                await requestContext.SendError(ex.Message);
+                await SendErrorAndCleanup(serializeParams.FilePath, requestContext, ex);
+            }
+        }
+
+        private async Task SendErrorAndCleanup(string filePath, RequestContext<SerializeDataResult> requestContext, Exception ex)
+        {
+            if (filePath != null)
+            {
+                try
+                {
+                    DataSerializer removed;
+                    inProgressSerializations.TryRemove(filePath, out removed);
+                }
+                catch
+                {
+                    // Do not care if there was an error removing this, must always delete if something failed
```

Comment:
>must always delete if something failed [](start = 72, length = 38)

You're not actually deleting the file though - is that intentional?

Reply:
Great question. I have a gap here for sure. On consideration, I will ensure the write stream is closed out in this case and intentionally leave file contents that have been written as-is. The thinking is that for most formats (CSV, JSON) users are going to be happier with half the written contents on an unexpected error than none.